warehouse/query-core/src/main/java/datawave/query/jexl/visitors/DelayedNonEventSubTreeVisitor.java

```diff
@@ -0,0 +1,132 @@
+ */
+public class DelayedNonEventSubTreeVisitor extends BaseVisitor {
+    private IteratorBuildingVisitor iteratorBuildingVisitor;
+    private Set<String> nonEventFields;
+    private Multimap<String,JexlNode> delayedNonEventFieldMapNodes;
+    
+    public static Multimap<String,JexlNode> getDelayedNonEventFieldMap(IteratorBuildingVisitor iteratorBuildingVisitor, ASTJexlScript script,
+                    Set<String> nonEventFields) {
+        // create a safe copy to rip apart
+        ASTJexlScript copy = TreeFlatteningRebuildingVisitor.flatten(script);
+        
+        // run the visitor on the copy
+        DelayedNonEventSubTreeVisitor visitor = new DelayedNonEventSubTreeVisitor(iteratorBuildingVisitor, nonEventFields);
+        copy.jjtAccept(visitor, null);
+        
+        return visitor.delayedNonEventFieldMapNodes;
+    }
+    
+    private DelayedNonEventSubTreeVisitor(IteratorBuildingVisitor iteratorBuildingVisitor, Set<String> nonEventFields) {
+        this.iteratorBuildingVisitor = iteratorBuildingVisitor;
+        this.nonEventFields = nonEventFields;
+        delayedNonEventFieldMapNodes = HashMultimap.create();
+    }
+    
+    /**
+     * Check if the node is an instance of a DelayedPredicate and if so capture all nonEvent delayed nodes, ranges, and node operators into
+     * delayedNonEventFieldMapNodes.
+     * 
+     * @param node
+     * @param data
+     * @return
+     */
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        if (ASTDelayedPredicate.instanceOf(node) || (data != null && (boolean) data)) {
+            // strip off only the delayed predicate, leave any other markers intact since they may be necessary to properly process inside the
+            // IteratorBuildingVisitor
+            JexlNode candidate = QueryPropertyMarker.getQueryPropertySource(node, ASTDelayedPredicate.class);
+            
+            if (candidate == null) {
+                // we are in a recursive situation inside an delayed predicate, evaluate the node
+                candidate = node;
+            }
+            
+            extractDelayedFields(candidate);
+            
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            return super.visit(candidate, true);
+        }
+        
+        // it wasn't a delayed node, process down the tree
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (data != null && (boolean) data) {
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
```

Comment:
I thought we already fixed this in the AndIterator/OrIterator ?  hmm..

Reply:
This is a result of us reusing the IteratorBuildingVisitor to avoid code duplication, and the rules for a root there require includes to be non-empty. I will update this documentation since it is not really a bug, but a limitation of the IteratorBuildingVisitor. Future refactoring could pull the functionality out.